Add documentation on v0 symbol mangling.

ehuss opened the pull request:
This adds official documentation for the v0 symbol mangling format, migrating the documentation from [RFC 2603](https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html).
The format was originally stabilized as the `-C symbol-mangling-version` option, but the specifics were not stabilized (per https://github.com/rust-lang/rust/pull/90128#issuecomment-948569123).
Per the discussion at https://github.com/rust-lang/rust/pull/93661#discussion_r799783363 this adds those specifics as an official description of the format.

cc #89917

ehuss committed:
Add documentation on v0 symbol mangling.

rust-highfive commented:
r? @GuillaumeGomez

(rust-highfive has picked a reviewer for you, use r? to override)

ehuss commented:
Before this is merged, there are a variety of questions to resolve:

* [ ] Is the compiler team ready for this?
* [ ] Are there any open questions about the format?
* [ ] How is it intended to extend this format in the future? For example, if additional *const* types are stabilized, can v0 just be extended to add them? What constitutes a need for a v1 format? Which parts of v0 can be extended at will?
* [ ] This currently only documents stable parts of the language. Should unstable things like additional const generic types be included now, or just when they are stabilized?
* [ ] The base62 encoding scheme subtly differs from common base62 (or any baseNN) encoding scheme by swapping the uppercase/lowercase letters. Was that intended?
* [ ] I can't figure out a way to create a symbol with a placeholder (`p`) const. How can one be created?
* [ ] I left a FIXME in the text for wasm-specific behavior. I'm not sure if that should be included?

Deviations from the RFC:

* [ ] I defined `<bytes>` for identifier as UTF-8. Should it be defined as arbitrary bytes instead? `rustc_demangle` for example does not support non-utf-8.
* [ ] I specified a 64-bit upper-bound for `<base-62-number>` so that demanglers can have a well-defined limit (and that is how rustc-demangle works). Should I keep that, or should demanglers be prepared for larger values?
* [ ] I added more detail on the encoding, where the RFC was under-specified. I did my best to make it correct but not over- or under-specified, though it wouldn't hurt for a careful review. For example:
    * fn ABI string converts dashes to underscores.
    * base62 encoding
    * The format for `<decimal-number>`.
* [ ] I added more detailed recommended demangling (based on rustc-demangle).

cc @eddyb @Mark-Simulacrum

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6662529455?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/8143f5ec2ad47756d44748af391267f5f25fa86f/checks/6662529455/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 9.635 seconds
doc tests for: /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md" "--test-args" ""

stdout ----

running 15 tests
running 15 tests
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Vendor_specific_suffix (line 1015) ... FAILED
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Lifetime (line 609) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Backref (line 931) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Nested_path (line 384) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Impl (line 286) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Crate_root (line 180) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Trait_impl (line 247) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Instantiating_crate (line 976) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Inherent_impl (line 213) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Trait_definition (line 332) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_path::Path__Generic_arguments (line 440) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Const (line 676) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Identifier::Punycode_identifiers (line 526) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Type (line 862) ... ok
test /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Symbol_name (line 114) ... ok
failures:


---- /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Vendor_specific_suffix (line 1015) stdout ----
error[E0412]: cannot find type `RefCell` in this scope
 --> /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md:1017:25
  |
4 |     pub static EXAMPLE: RefCell<u32> = RefCell::new(1);
  |
help: consider importing this struct
  |
2 | use std::cell::RefCell;
2 | use std::cell::RefCell;
  |

error[E0433]: failed to resolve: use of undeclared type `RefCell`
 --> /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md:1017:40
  |
4 |     pub static EXAMPLE: RefCell<u32> = RefCell::new(1);
  |
help: consider importing this struct
  |
2 | use std::cell::RefCell;
---
For more information about an error, try `rustc --explain E0412`.
Couldn't compile the test.

failures:
    /checkout/src/doc/rustc/src/codegen-options/symbol-mangling.md - Symbol_Mangling::v0_mangling_format::Vendor_specific_suffix (line 1015)
test result: FAILED. 14 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.23s


stderr ----
```

</details>

eddyb commented:
I'm going to try to do a point-by-point reply to https://github.com/rust-lang/rust/pull/97571#issuecomment-1141551210 (but some of this feels like it's going over things discussed before the RFC was accepted? I suppose those things tend to get lost to time)

* > Is the compiler team ready for this?

  Not speaking for anyone else but I've been operating under the assumption that the RFC itself is the documentation (though it seems reasonable for it to be accessible through the reference, not sure why we didn't just do that), and the main thing we've been waiting on is upstreaming into non-Rust tooling.

* > Are there any open questions about the format?

  Can't think of anything other than the discussion in https://github.com/rust-lang/rfcs/pull/3161 (which hasn't been merged so I assume it's not part of this PR)

* > How is it intended to extend this format in the future? For example, if additional *const* types are stabilized, can v0 just be extended to add them? What constitutes a need for a v1 format? Which parts of v0 can be extended at will?

  Generally, as long as leave gaps in the grammar (i.e. unused "letters" in each of 3-4 grammar categories) we can always extend it - by a quick count we have:
  * 7 for path/type (uppercase): `GHJLUVW`
  * 5 for type (lowercase): `gkqrw`
  * 23 for "well-known" namespace (uppercase): everything except `C` (closure), `S` (shim) and `I` (impl, but only in [a specific polymorphization situation](https://github.com/rust-lang/rust/blob/989b806f61cf12840ddcd8b2bf16d6bef055bd01/compiler/rustc_symbol_mangling/src/v0.rs#L321-L334)).
  * 21 for implementation-defined namespace (lowercase): this one doesn't really matter, since we just assign them in the compiler and the demangler doesn't have to know about them - worst case we can start putting information into the low bits of the disambiguator if we ever run out

  For constants, https://github.com/rust-lang/rfcs/pull/3161 both supports the vast majority of types that we *could ever* have in `const` generics, *and* offers (in last paragraph of the PR description) a *generalized* fallback we could *always* do in other cases (without updating demanglers at all).

  As for v1, that would require backwards-incompatible changes, i.e. repurposing characters in ways that could accidentally be incorrectly decoded as *unrelated* v0 manglings. If I had to guess, this would likely a binary format, usable on platforms that do not have the charset limitations, and may even include some kind of compression using a standardized dictionary (zstd was mentioned during the original v0 RFC).

* > This currently only documents stable parts of the language. Should unstable things like additional const generic types be included now, or just when they are stabilized?

  https://github.com/rust-lang/rfcs/pull/3161 isn't even accepted yet so I don't think there's much of a point in including it *now*. OTOH, `rustc-demangle` does implement it, so it might be more of a matter of what we do by default?

* > The base62 encoding scheme subtly differs from common base62 (or any baseNN) encoding scheme by swapping the uppercase/lowercase letters. Was that intended?

  Not sure, I think that was all chosen by @michaelwoerister before I got involved, and I kept it. Though I would've naively assumed lowercase goes before uppercase in any base, so I wouldn't trust myself to have gotten it right either. <sub>(Actually, doesn't e.g. base64 look nothing like hex and instead makes `A` the lowest value digit, so that a bunch of zero bytes turn into `AAAAA...` in base64? So it's not just the order of lowercase and uppercase)</sub>

* > I can't figure out a way to create a symbol with a placeholder (`p`) const. How can one be created?

  It should be the same as with the type placeholder: `impl` generics that do not actually apply to the entity being described, like a `static`, which is never monomorphized by parameters in scope (e.g. [mangled](https://godbolt.org/z/8ddosWq3E) / [demangled](https://godbolt.org/z/s8M8e7Tca) - `I...pKpE` is `...<_, _>`, a type and a const placeholder).

* > I left a FIXME in the text for wasm-specific behavior. I'm not sure if that should be included?

  Maybe, in the general sense of linker symbol documentation, since wasm has a somewhat unique two-level system, but it wouldn't be relevant to a v0 explanation (which IMO should have a clearly marked document of its own, not mixed with other linker symbol concerns like `#[no_mangle]` and whatnot).

* > Should there be a more detailed explanation of the mapping of how Rust entities are encoded?

  I don't think so, the format isn't designed to be reproducible between two producers, it's more of a "heterogeneous syntax serialization" (honestly we could've probably deduplicated more, or chosen better letters if more time was spent on that front alone, but it's not too bad).

  > Deviations from the RFC:

* > I defined `<bytes>` for identifier as UTF-8. Should it be defined as arbitrary bytes instead? `rustc_demangle` for example does not support non-utf-8.

  The whole charset is `_0-9a-zA-Z` (ident-wise, the subset of `XID_Start`+`XID_Continue` that overlaps ASCII) so I think `<bytes>` was just used as a shorthand for *that* - in theory, it could be UTF-8, but punycode was added specifically so that we could have Unicode w/o UTF-8.

* > I specified a 64-bit upper-bound for `<base-62-number>` so that demanglers can have a well-defined limit (and that is how rustc-demangle works). Should I keep that, or should demanglers be prepared for larger values?

  Demanglers can have their own implementation limits, I don't think we should put that in the format.
  E.g. `rustc-demangle` also has like a `[char; 128]` limit for punycode (because it's `#![no_std]` so it can't allocate a `Vec<char>`), and demangling still succeeds, but that overlong identifier is printed out in the standard punycode syntax (used by DNS, primarily).

  For encoding constants I went out of my way to allow "just dump the mangled form" by having it be hex so it just needs a `0x` prefix printed (at the cost of compactness compared to base62).

  Honestly, the only reason I can think for disambiguators using base62 is the crate disambiguator hash, and even then, unless it's actually printed, fully decoding it into a number shouldn't be necessary (so `rustc-demangle` could be updated to make the limit "softer").

* > I added more detail on the encoding, where the RFC was under-specified. I did my best to make it correct but not over- or under-specified, though it wouldn't hurt for a careful review. For example:
  >  * fn ABI string converts dashes to underscores.
  >  * base62 encoding
  >  * The format for `<decimal-number>`.
* > I added more detailed recommended demangling (based on rustc-demangle).

  For both of these, I'm tempted to say the original RFC should be amended, since more features aren't added, just existing information corrected - there is however [some discussion](https://github.com/rust-lang/rfcs/pull/3224#issuecomment-1053323145) on an existing example of such an amendment so maybe I'm wrong.

ehuss committed:
Rearrange symbol-mangling chapter out of codegen-options.

ehuss committed:
Remove 64-bit limit for base-62-numbers.

Demanglers should be prepared for any arbitrary length number.

ehuss commented:
Thank you for the detailed response!

> Shouldn't this go in the reference? And then have the CLI option link to that?

Hm, my impression was the opposite, that this is not a part of the language and just an implementation-specific concern. The RFC was accepted by the compiler team, not the lang team. And it isn't really clear what consequence it has for the language if it isn't to be used as a stable ABI or have well-defined translation from Rust entities.

I can imagine in the future that this could potentially provide the basis for a stable ABI, but that seems very far out, no?

>  v0 explanation (which IMO should have a clearly marked document of its own, not mixed with other linker symbol concerns like #[no_mangle] and whatnot)

Yea, I waffled a bit on how to organize it.  I decided to move it to a separate chapter.

> Demanglers can have their own implementation limits, I don't think we should put that in the format.

OK, I removed the limit.

> Generally, as long as leave gaps in the grammar (i.e. unused "letters" in each of 3-4 grammar categories) we can always extend it 

I can understand how something like `namespace` or `basic-type` could be extended, but if any of the other open tags were extended, a demangler would have no way to know how to interpret anything that follows. How would that be different from any other backwards-incompatible change? The end result is that an older demangler would be unable to demangle the symbol.

fbstj commented on src/doc/rustc/src/symbol-mangling/index.md:
````diff
@@ -0,0 +1,52 @@
+```
+
+[`rustc-demangle`]: https://crates.io/crates/rustc-demangle
+[`rustfilt`]: https://crates.io/crates/rustfilt
````
is recommending tools via crates.io the norm?

fbstj replied:
ahh well yeah looking at that page in-particular it seems like they "recommend" using `cargo install` rather than having binaries/installers. I was specifically meaning _tools_ rather than crates / libraries but yeah this crates pages is probably the most sensible page for this tool

michaelwoerister commented:
Wow, that is some extensive documentation 😀
I'll take a closer look next week.

> The base62 encoding scheme subtly differs from common base62 (or any baseNN) encoding scheme by swapping the uppercase/lowercase letters. Was that intended?

Nope, there is no deeper meaning to that. Would have been nice to catch that during RFC review, but I guess now that there are demanglers in external tools, it's too late to change it.

eddyb commented:
* > > Shouldn't this go in the reference? And then have the CLI option link to that?
  >
  > Hm, my impression was the opposite, that this is not a part of the language and just an implementation-specific concern. The RFC was accepted by the compiler team, not the lang team. And it isn't really clear what consequence it has for the language if it isn't to be used as a stable ABI or have well-defined translation from Rust entities.

  The exact mapping of source code to symbols is implementation-specific, but any Rust implementation would immediately benefit from generating v0 symbols, as tools exist to demangle them back into *Rust syntax*, while the format also is able to hide enough redundancy to cover various additional sources of "identity" not explicitly covered otherwise.
  
  I agree that in retrospect the RFC being strictly compiler team is a bit weird, might be a good idea to poll @rust-lang/lang on how they feel about treating the v0 symbol mangling the way I described ("officially a Rust mangling scheme").

* > I can imagine in the future that this could potentially provide the basis for a stable ABI, but that seems very far out, no?

  Probably only in the vague sense that it's the first sketch we have of being able to serialize what we might consider "definition identity". Definitely agree on far out.
  
  Earliest similar thing I can think of would be #95845, if we stabilized a method on `TypeId` to actually get the mangled string (or more likely something that can be *decompressed* into one), but that PR is not looking favorable right now so that might never happen.

* > > Generally, as long as leave gaps in the grammar (i.e. unused "letters" in each of 3-4 grammar categories) we can always extend it 
  >
  > I can understand how something like `namespace` or `basic-type` could be extended, but if any of the other open tags were extended, a demangler would have no way to know how to interpret anything that follows. How would that be different from any other backwards-incompatible change? The end result is that an older demangler would be unable to demangle the symbol.

  Ah I see now. I think there's two levels of incompatibility (not sure how to describe each):
  * symbols using some new feature require new (e.g. type) encoding in mangling
    * only affects code using the new feature, not existing code
    * getting mangling support out there can be part of stabilization
    * we do have the fallback option of choosing a less-appealing-after-demangling encoding, because any syntactic category that can embed paths can represent arbitrary trees
      * e.g. `A<B<C, D>, X<Y>>` is no different from the sexpr `(A (B C D) (X Y))`, information-wise
  * new symbols repurpose existing mangling syntax to mean something else
    * this is worse because demanglers can then get confused and "succeed" with the wrong output
    * IOW, this introduces ambiguity, *unless* some kind of "explicit version bump" is done

michaelwoerister commented on src/doc/rustc/src/symbol-mangling/v0.md:
```diff
@@ -0,0 +1,1172 @@
+> inherent-impl → `M` *[impl-path]* *[type]*
+
+An *inherent-impl* indicates a path to an [inherent implementation][reference-inherent-impl].
+It consists of the character `M` followed by an *[impl-path]* to the impl's parent followed by the *[type]* representing the `Self` type of the impl.
```
I think this should be something like
> It consists of the character `M` followed by an *[impl-path]*, which uniquely identifies the `impl` block the item is defined in.

impl-path is the parent of the impl plus the index of the impl within the parent. We just don't emit the index (disambiguator) if it's equal to zero.

michaelwoerister commented on src/doc/rustc/src/symbol-mangling/v0.md:
```diff
@@ -0,0 +1,1172 @@
+>     │└─────────┬──────────┘└────┬──────┘
+>     │          │                │
+>     │          │                └── Self type "Example"
+>     │          └─────────────────── path to the impl's parent "mycrate"
```
This example should probably contain a disambiguator for the impl-path to make it more representative. Most impl-paths will have one.

michaelwoerister commented on src/doc/rustc/src/symbol-mangling/v0.md:
```diff
@@ -0,0 +1,1172 @@
+This helps differentiate symbols that would otherwise be identical,
+for example the monomorphization of a function from an external crate may result in a duplicate if another crate is also instantiating the same generic function with the same types.
+
+In practice, the instantiating crate is also the crate where the symbol is defined,
```
```suggestion
In practice, the instantiating crate is also often the crate where the symbol is defined,
```

michaelwoerister commented:
Great docs, @ehuss! I left some comments in a few places but that's mostly nit-picking.

ehuss committed:
Update from review from michaelwoerister.

ehuss commented on src/doc/rustc/src/symbol-mangling/v0.md:
```diff
@@ -0,0 +1,1172 @@
+followed by a *[path]* which is a path representing the parent of the entity,
+followed by an *[identifier]* of the entity.
+
+The identifier of the entity may be empty when the entity is not named.
```
I have tweaked the wording to make try to make it clearer. I wasn't trying to say that there is *no* identifier, just that its length is zero.

eddyb commented:
@michaelwoerister Do any of these comments apply to anything taken from the RFC text? It would be nice to fix them there (since IMO a diff of the RFC and reference should mostly have large-scale reorganization and additions, not e.g. spelling fixes in copied text).

tmiasko commented on src/doc/rustc/src/symbol-mangling/v0.md:
```diff
@@ -0,0 +1,1175 @@
+> A *lifetime* may be displayed like a Rust lifetime using a single quote.
+> Index 0 should be displayed as `'_`.
+>
+> Lifetimes starting from 1 may be translated to single lowercase letters starting with `'a`.
+> Indices over 25 may consider printing the numeric lifetime index as in `_123`.
+>
+> Index 0 should not be displayed for lifetimes in a *[ref-type]*, *[mut-ref-type]*, or *[dyn-trait-type]*.
+>
+> Nested binders may consider tracking their indices so that lifetime lettering can start back with `'a` within a nested binder.
+> See *[binder]* for more on lifetime indexes and ordering.
```
A lifetime can be displayed by converting De Bruijn index to De Bruijn level (level = number of bound lifetimes - index) and selecting a unique name for each level. For example, starting with `'a` for level 0.

ehuss replied:
Thanks, I rewrote this trying to incorporate your suggestion.

ehuss committed:
Clarify missing tick.

ehuss committed:
Clarify grammar for decimal-number cannot have leading zeroes.

ehuss committed:
Rewrite recommended demangling for lifetimes using "De Bruijn level".

michaelwoerister commented:
> @michaelwoerister Do any of these comments apply to anything taken from the RFC text? It would be nice to fix them there (since IMO a diff of the RFC and reference should mostly have large-scale reorganization and additions, not e.g. spelling fixes in copied text).

No, I don't think so.

Dylan-DPC commented:
@michaelwoerister any updates on this?

michaelwoerister commented:
Not really. I think we might want to try and define which of the questions in https://github.com/rust-lang/rust/pull/97571#issuecomment-1141551210 really need to be answered before merging this PR and which can be left unanswered for now. E.g. "How is it intended to extend this format in the future?" is an open question (I think) that will need some careful investigation.

michaelwoerister commented:
> * I defined `<bytes>` for identifier as UTF-8. Should it be defined as arbitrary bytes instead? `rustc_demangle` for example does not support non-utf-8.

The RFC says "[the mangling scheme] generates symbols that only consist of the characters `A-Z`, `a-z`, `0-9`, and `_`". However, some common platforms have no trouble with UTF-8 symbols, in which case the compiler can emit identifiers as UTF-8 instead of as Punycode. Here's the relevant section in the RFC: https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html#punycode-vs-utf-8

In my opinion this boils down to a demangler having to support both UTF-8 and Punycode here, but not arbitrary bytes.

michaelwoerister commented on src/doc/rustc/src/symbol-mangling/v0.md:
```diff
@@ -0,0 +1,1182 @@
+
+- It provides an unambiguous string encoding for everything that can end up in a binary's symbol table.
+- It encodes information about generic parameters in a reversible way.
+- The mangled symbols are *decodable* such that demangled form should be easily identifiable as some concrete instance of e.g. a polymorphic function.
```
```suggestion
- The mangled symbols are *decodable* such that the demangled form should be easily identifiable as some concrete instance of e.g. a polymorphic function.
```

Dylan-DPC commented:
@ehuss any updates on this?

ehuss committed:
Add an example of placeholders.

ehuss committed:
Add missing word "the".

ehuss commented:
Sorry, I kinda let this languish.

Summarizing my thoughts on the open questions:

> Is the compiler team ready for this?

Only they can answer this. I would say that the format is de facto public since it is available in various external tooling, so I think it should be documented and committed to as somewhat stable.

> Shouldn't this go in the reference?

I don't have a strong preference on one place or the other. My thinking was that this is mostly an implementation detail, and that's why I started here, and also that's what Mark recommended. I also expected it to be easier to land. I think it is a bigger question for Rust The Language if it should prescribe a mangling scheme. I'm also a little hesitant since this doesn't define how to mangle a Rust entity (it only defines how to interpret an already mangled symbol). However, I see the value in saying Rust implementations should use this format to retain compatibility between them.

I'm fine with either option, though getting it into the reference might be harder since we'll need to get the lang team involved.

> Are there any open questions about the format?

Other than extensions, I'm not aware of any.

> How is it intended to extend this format in the future? 

This is a big question that I think deserves a section in the documentation. My suggestion is something like this:

> ## Extensions
> 
> This format may be extended in the future to add new tags as Rust is extended with new language items. To be forward compatible, demanglers should gracefully handle symbols that have encodings where it encounters a tag character not described in this document. For example, they may fall back to displaying the mangled symbol. The format may be extended anywhere there is a tag character, such as the *[type]* rule. The meaning of existing tags and encodings will not be changed.

which I think is pretty close to what eddyb said. I don't think we should freeze the v0 format, and then roll out new versions for minor extensions, as that sounds like quite a hassle with limited benefit. I can see how changing the meaning of something could confuse a demangler and cause it to demangle something incorrectly.

> FIXME about wasm `#[link(wasm_import_module="…")]` behavior.

I don't know wasm well enough to say exactly how this should be documented. I'm fine with leaving it to a future PR. However, if someone wants to suggest the text to add, I'd be happy to add it. I don't know if this is LLVM-specific, if it should be documented, or if it is intended to change.

> Clarifications of the RFC, that need a more careful review.
> Adjusted the grammar for decimal-number to make it clear it cannot have leading zeroes.

I'm not sure if anyone reviewed the clarifications and fixes that I added. They are listed in the PR description. Happy to answer any questions. I'm fairly confident they are correct, but a second set of eyes would be good.

> I added more detailed recommended demangling (based on rustc-demangle).

I personally like the recommendations, so I would like to keep them.

----

I think that's it, so I consider this ready for review again.

@rustbot ready

michaelwoerister commented:
> > Extensions
This format may be extended in the future to add new tags as Rust is extended with new language items. To be forward compatible, demanglers should gracefully handle symbols that have encodings where it encounters a tag character not described in this document. For example, they may fall back to displaying the mangled symbol. The format may be extended anywhere there is a tag character, such as the [type] rule. The meaning of existing tags and encodings will not be changed.
>
> which I think is pretty close to what eddyb said. I don't think we should freeze the v0 format, and then roll out new versions for minor extensions, as that sounds like quite a hassle with limited benefit. I can see how changing the meaning of something could confuse a demangler and cause it to demangle something incorrectly.

That's my general opinion too. I've been thinking about forward compatibility and graceful degradation for demanglers a bit lately. I think that it _would_ be possible to enable more graceful degradation than we currently have (by changing the concrete encoding while leaving the abstract syntax as it is). But given how hard it is to push updates through all the downstream tooling and that the benefit would be mostly cosmetic and only make a difference for the few symbols that contain unknown syntax, I don't think it's worth the trouble.

ehuss committed:
Add description of forwards-compatible behavior.

michaelwoerister commented:
OK, I propose that we merge this. It is a great piece of documentation and as @pnkfelix noted on Zulip, we should not keep it from being available any longer just because there are some open questions (and most of them have been [answered](https://github.com/rust-lang/rust/pull/97571#issuecomment-1546799329) meanwhile). Thanks for all the work you put into this, @ehuss!

@rfcbot fcp merge

ehuss committed:
Fix span for punnycode

ehuss commented on src/doc/rustc/src/symbol-mangling/v0.md:
````diff
@@ -0,0 +1,1222 @@
+
+```text
+_RNvNtNtCsgOH4LzxkuMq_7mycrateu8gdel_5qa6escher4bach
+                              ││└───┬───┘
````
Thanks for spotting that!

michaelwoerister commented:
@ehuss, feel free to r=me when you think it's ready.

ehuss commented:
Thanks for the reviews!

@bors r=michaelwoerister rollup

bors commented:
:pushpin: Commit da4f62ebf7c487700c21a713c0cc3974bef11fc8 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister da4f62ebf7c487700c21a713c0cc3974bef11fc8 -->
<!-- homu: {"type":"Approved","sha":"da4f62ebf7c487700c21a713c0cc3974bef11fc8","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
I tried to add this to a rollup but somehow it "slipped through" and didn't get actually merged in. I am assuming that this should be marked
@bors rollup=never

bors commented:
:hourglass: Testing commit da4f62ebf7c487700c21a713c0cc3974bef11fc8 with merge 0ca28c3da75b165cc3e4870a1f397a23d2a58857...
<!-- homu: {"type":"BuildStarted","head_sha":"da4f62ebf7c487700c21a713c0cc3974bef11fc8","merge_sha":"0ca28c3da75b165cc3e4870a1f397a23d2a58857"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5687809758/job/15418181681)
Approved by: michaelwoerister
Pushing 0ca28c3da75b165cc3e4870a1f397a23d2a58857 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5687809758/job/15418181681"},"merge_sha":"0ca28c3da75b165cc3e4870a1f397a23d2a58857"} -->

rust-timer commented:
Finished benchmarking commit (0ca28c3da75b165cc3e4870a1f397a23d2a58857): [comparison URL](https://perf.rust-lang.org/compare.html?start=0699d9951695c0d8ad2efa1e208e97bcc5dfbb53&end=0ca28c3da75b165cc3e4870a1f397a23d2a58857&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0699d9951695c0d8ad2efa1e208e97bcc5dfbb53&end=0ca28c3da75b165cc3e4870a1f397a23d2a58857&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0699d9951695c0d8ad2efa1e208e97bcc5dfbb53&end=0ca28c3da75b165cc3e4870a1f397a23d2a58857&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.6%, 0.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.3%, 2.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-3.5%, -2.0%] | 2     |
| All ❌✅ (primary)                 | 0.6%  | [0.6%, 0.6%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0699d9951695c0d8ad2efa1e208e97bcc5dfbb53&end=0ca28c3da75b165cc3e4870a1f397a23d2a58857&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=0699d9951695c0d8ad2efa1e208e97bcc5dfbb53&end=0ca28c3da75b165cc3e4870a1f397a23d2a58857&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 653.092s -> 653.16s (0.01%)
